### ISSUE TITLE
Roadmap 2026

### ISSUE THREAD

**Issue by fkiraly** [sktime/pytorch-forecasting#1993]
From roadmap planning meeting on 2025-11-10

Present: @agobbifbk, @ankurankan, @fkiraly, @julian-fong, @phoeenniixx, @PranavBhatP, @RecreationalMath, @xaheli

### Topics & Prioritization

**pytorch-forecasting v2 API rework** :+1: :+1: :+1::+1: :heavy_check_mark: 
* Addition of `predict` function and full integration of `pkg` class with rest of D1,D2, model layer pipeline
* Addition of preprocessing steps - scaling, transformations etc in D1/D2 layers
* Improve the test coverage. 
* Add temporal splitting
* Migrate the models from v1 to v2 and deprecate v1
* Add extension Templates for new contributors.
* Add adpaters for `nn` losses.

**Extending model zoo** :+1::+1::+1::heavy_check_mark:
* Addition of new models to DSIPTS.

**Maintenance items, CI, tests, avoiding tech debt** :+1::+1: :heavy_check_mark:
* initial review of `pytorch-tabular` API, comparison to `pytorch-forecasting`
* Improve the test coverage in `pytorch-forecasting`
* negative tests and validation tests to be added for PyTorch based models. 
* Add python-3.14 support for ptf (scheduled for the next release).
* urgent maintenance items in `pytorch-tabular` - `pyproject`, python 3.14
* adding `torch` to all extras in `sktime`, or replace `tensorflow` by `torch`

**backend - `sktime` GPU support for `torch` based models** :+1:

* Support for GPU utilisation to be added

**API homogenization of `torch` models within `sktime`** :+1::+1::+1::heavy_check_mark:
* Unify scattered `torch` dataset classes for classification, regression etc.
* multiple `_pytorch.py` exist in sktime repo, they should all be unified into a single class in a single file with a `_pytorch_util.py` containing all the helper functions.
* activation, criterion, optimizer, callbacks initializations currently done crudely. Improvements to be made.

**API homogenization between `sktime`, `pytorch-forecasting`, `pytorch-tabular`, DSIPTS** :+1: :+1: :+1: :+1::heavy_check_mark::heavy_check_mark:

* uniform way of network creation and definition. API design that can be re-used everywhere.
* Create an interface for `pytorch-tabular` and `pytroch-forecasting` much like `prophetverse` and `sktime`
* Having a uniform way of implementing DL based models across sktime, Ptf & ptt. Since we are implementing PyTorch based models in sktime from scratch, it will be great if we can borrow best practices from sister libraries and implement them in sktime. 
* uniform way of network creation and definition. API design that can be re-used everywhere (sktime, ptf, ptt). This would also help in maintaining all three repos and allow cross collaboration.
* Handling training and prediction pipelines in a more modular way than currently in sktime. Learnings from ptf & ptt to be incorporated.
* Synchronise on ideas/approaches on Data layer & Modelling; ensuring benefit to both the repos.
* Enhancement related to common workflows in D1-D2 layers. 
    

### New Features

**Global model support in PTF and sktime**
* adding groups (categroical variable) and distinction between global and no global forecast

**Foundation model support in PTF** :+1::+1: :heavy_check_mark: 
* Addition of foundation models to PTF v2 (in parallel with the rework) - helps confirm the feasibility of the design for all the proposed models for v2. Re: [open issue](https://github.com/sktime/pytorch-forecasting/issues/1959)

**Benchmarking support**

* Benchmarking support for PTF v2. Very open ended right now.

### Workstreams

one big workstream due to interconnection

* design / API makes most sense jointly
    * have design for `pytorch-forecasting`
    * no consilated design for `sktime` / `torch`
    * "inherited" design for `pytorch-tabular`
* short-term, "big projects"
    * maintenance items are independent
    * `pytorch-forecasting` v2
    * `pytorch-forecasting` adding models - contingent on API work
    * `sktime` API refaector
    * `sktime` adding models (could happen right away)
    * `pytorch-tabular` (anything) - need to onboard maintainers to `pytorch-tabular`
    * cross-cutting design work
    * cross-cutting API changes (contingent on design)

Prioritization:

* progress work that is currently ongoing
    * `pytorch-forecasting` v2 and model add
    * `sktime` API rework
    * `pytorch-tabular` basic maintenance
        * perhaps onboard new mentees
    * DSIPTS rework

* design review
    * create shared design document on the topic of `torch` related API in 4 packages
        * action Nirbhai - new hackmd doc to fill - collect current designs and notes
        * action Aryan - send link to existing ptf document
    * presentations where necessary (tbd)
    * aim is for all to gain working knowledge of the three packages
    * based on that, joint design work

**Comment by Spinachboul** [comment]
@fkiraly 
Hi Franz! I would like to add a suggestion here.
Why dont we also include **explicitly** the ownwrship rules and consequence which can make things more clear?
Here is an example format which might sound sense -----
## Pytorch Forecasting
**Role**: Deep learning forecasting engine
**Status**: Source of truth for DL forecasting

### PTF owns:
- Temporal Windowing
- Temporal Splitting for training/validation
- Forecast Horizon Handling
- Scaling, Normalization, Categorical Encoding
- Dataset Semantics for DL forecasting
- Dataset Loaders and batching
- Training Loops
- GPU Utilization
- Prediction Semantics for DL models

### PTF Must Expose:
- A complete `fit` + `predict` interface
- Outputs that can be wrapped without reinterpreting semanrtics
- Preprocessing and Splitting can be configured

## Sktime
**Role**: User-facing time series framework and orchestration
**Status**: Consumer of engines, not an engine itself

### Sktime owns
- Unified API (`fit`, `predict`, `update`)
- Forecasting Horizon absraction
- Pipeline and Transformers
- Cross-Validation and benchmarking

> Similarly we can define the ownership for PTT and  DSIPTS
@phoeenniixx  and @PranavBhatP  your inputs here are valuable....

**Comment by fkiraly** [comment]
@Spinachboul, I think architecturally it is already quite clear where the scope distinction lies between `sktime` and `pytorch-forecasting` - the latter covers exactly everything that is `torch` specific.

You can read the following documents and discussions to get up to speed:
https://github.com/sktime/pytorch-forecasting/issues/1736
https://github.com/sktime/enhancement-proposals/pull/39
https://github.com/sktime/enhancement-proposals/pull/46

**Comment by selcuksntrk** [comment]
Hi @phoeenniixx @agobbifbk @PranavBhatP ,

I'm Selcuk Senturk, an AI/ML Engineer at IBM Expert Labs and MSc Computer Science student at Ege University. I'm interested in contributing to the pytorch-forecasting v2 rework as part of GSoC 2026.

The areas I'd like to focus on: completing the D1/D2 and pkg layer work, adding train_test_split strategies, adding adapter for nn losses and MultiLoss support and improving the test framework to cover more training/validation scenarios and edge cases.

Relevant background: 
I work daily with PyTorch-based production systems at Siskon Software I built real-time deep learning pipelines for computer vision, and at IBM I handle model training, optimization, and deployment on Watsonx.ai. During my BSc research I also trained deep learning models for protein family classification from molecular structure, so I'm comfortable with the full training loop not just calling high-level APIs. I've also worked with multi-agent orchestration systems, which gives me some intuition for modular, layered architecture design like what v2 is aiming for.

I would like to discuss further if my background is adequate and how I plan to achieve these implementations.

Resume: [Drive](https://drive.google.com/file/d/12NCptLXY1FZIreGX6pGKuQFi17XRQjTQ/view?usp=sharing)
LinkedIn: [selcuksntrk](https://www.linkedin.com/in/selcuksntrk/)

**Comment by ShreyasN707** [comment]
Hi @fkiraly 
Building on the API homogenization discussion, one idea that might help across sktime / PTF / PTT is introducing a lightweight TrainingConfig / EngineConfig object that cleanly separates:

- Model architecture definition

- Training loop configuration (optimizer, scheduler, callbacks, precision, device)

- Data semantics (windowing, grouping, scaling)

Right now, a lot of these concerns are mixed across model constructors and D1/D2 layers. A shared configuration object (even if package-specific at first) could:

- Reduce duplication across repos

- Make cross-library adapters easier

- Help future foundation model integration (where training may be optional or inference-only)

- Improve testability by isolating training semantics from model definition

- This might also help when unifying dataset and model abstractions between sktime and PTF v2.

Curious whether something like this has already been considered in the v2 design discussions.

**Comment by XVX-016** [comment]
Hi, I'm Tanmmay, a GSoC 2026 applicant who has submitted a proposal to GC.OS covering two items from this roadmap: foundation model integration (specifically Chronos via a sktime-compatible wrapper) and GPU device-agnostic training for v2 models via PyTorch Lightning. I've also commented on issue #2051 which tracks the Chronos work specifically. Looking forward to contributing happy to receive feedback on my approach before the deadline.

**Comment by sanju234-san** [comment]
Hi! I'm **Sanjeevni Dhir**, a GSoC 2026 applicant to GC.OS. I've submitted a proposal covering 
all three foundation model adapters — Chronos, Moirai, and TimesFM — into the v2 BaseForecaster 
interface, and I've just opened a draft PR here: [https://github.com/sktime/pytorch-forecasting/pull/2242]

Following @Spinachboul's suggestion to move to code rather than opinion — here's my full 
technical approach before I fill in the implementation.

---

## 🎯 What I'm Building

The core problem this issue identifies is real: pytorch-forecasting's D2 layer and foundation 
models speak completely different languages. My proposal bridges that gap with three thin 
**Adapter-pattern wrappers**:
```
User code (unchanged)
       │
       ▼
┌─────────────────────────────────────────┐
│         pytorch-forecasting v2          │
│                                         │
│  model = ChronosForecaster(...)         │
│  model.predict(dataloader)  ◄────────── │─── same API as TFT / N-BEATS
│  model.backtest(dataset)                │
└────────────────┬────────────────────────┘
                 │  adapter translates here
                 ▼
┌─────────────────────────────────────────┐
│         Foundation Model Layer          │
│                                         │
│  ChronosPipeline.predict()              │
│  MoiraiForecast.forward()               │
│  TimesFm.forecast()                     │
└─────────────────────────────────────────┘
```

---

## 🔄 The Core Translation Problem (and my solution)

The exact mismatch @Spinachboul described — PF's D2 batch format vs Chronos's expected input:
```
PF TimeSeriesDataSet batch          Chronos expected input
─────────────────────────           ──────────────────────
{                                   [
  "encoder_cat": Tensor,              {
  "encoder_cont": Tensor,               "target": Tensor,        
  "decoder_cat": Tensor,                "past_covariates": Tensor,
  "decoder_cont": Tensor,               "future_covariates": Tensor
  "encoder_lengths": Tensor,          }
  "decoder_lengths": Tensor,        ]
  "encoder_target": Tensor,
}
```

My adapter handles this translation **internally** — invisible to the caller:
```python
def _convert_batch(self, batch: dict) -> list[dict]:
    """
    Converts PF D2 batch format → Chronos input format.
    Handles context-length truncation and short-series padding.
    """
    return [
        {
            "target": batch["encoder_target"][i],
            "past_covariates": batch["encoder_cont"][i],
            "future_covariates": batch["decoder_cont"][i],
        }
        for i in range(len(batch["encoder_target"]))
    ]
```

---

## 🏗️ Architecture: Three Adapters, One Interface
```
BaseForecaster (v2)
        │
        ├── ChronosForecaster
        │       ├── _convert_batch() → Chronos input format
        │       ├── _run_inference() → ChronosPipeline.predict()
        │       └── _convert_output() → PF Prediction object
        │
        ├── MoiraiForecaster  
        │       ├── _convert_batch() → MoiraiForecast input format
        │       ├── _run_inference() → MoiraiForecast.forward()
        │       └── _convert_output() → PF Prediction object (quantiles native)
        │
        └── TimesFMForecaster
                ├── _convert_batch() → TimesFM input format
                ├── _run_inference() → TimesFm.forecast()
                ├── _convert_output() → PF Prediction object
                └── fine_tune() → optional dataset-specific adaptation
```

Each adapter exposes **identical methods** to TFT or N-BEATS:

| Method | Description |
|--------|-------------|
| `predict(dataloader)` | Zero-shot inference, returns `Prediction` |
| `predict_raw(dataloader)` | Raw model output before PF formatting |
| `backtest(dataset)` | Rolling-window evaluation on historical data |

---

## 📦 Handling the Three Models

| | Chronos | Moirai | TimesFM |
|---|---|---|---|
| **Org** | Amazon | Salesforce | Google |
| **Architecture** | T5 encoder-decoder | Transformer + patching | Decoder-only |
| **Input** | Univariate | Multivariate ✅ | Univariate |
| **Output** | Probabilistic samples | Native quantiles ✅ | Point + quantiles |
| **Licence** | Apache 2.0 ✅ | CC BY-NC 4.0 ⚠️ | Apache 2.0 ✅ |
| **Key dep** | `autogluon.timeseries` | `uni2ts` | `timesfm` |
| **Optional import?** | Yes | Yes | Yes |

> ⚠️ Moirai's CC BY-NC 4.0 licence will be clearly documented. The adapter code itself is Apache 2.0 — users are responsible for compliance with Moirai's licence terms.

---

## ❓ My Open Questions for Maintainers

Before I fill in the full implementation, I'd love guidance on:

1. **File location** — Should each adapter live in `pytorch_forecasting/models/chronos/`, or should all three live under a shared `pytorch_forecasting/models/foundation_models/` directory?

2. **Base class** — Should these subclass `BaseForecaster` directly, or is there a v2-specific inference-only base class more appropriate?

3. **Optional dependencies** — Preferred pattern for optional heavy deps: 
```python
   try:
       from chronos import ChronosPipeline
   except ImportError:
       raise ImportError("Install chronos: pip install autogluon.timeseries")
```
   Or separate extension packages entirely?

4. **D2 input mapping** — Has a preferred direction emerged between @Spinachboul's D2 subclassing approach and @agobbifbk's approach from the thread above?

---

Draft PR is open here for structural feedback: [https://github.com/sktime/pytorch-forecasting/pull/2242]  
Happy to adjust the approach based on maintainer input before filling in the implementation. 🙏

**Comment by agobbifbk** [comment]
Actually the idea of having converters at batch level is a good idea (near to the forward loop). 
In my opinion is better to not use directly `ChronosPipeline` or similar, when it is possible to create a batch adapter. What I've seen from Chronos is that the adapter will be straightforward if there are some checks in the D1/D2 layer: order of covariates, the future covariates must be also in the past, etc etc.  Can you please have a look to `TimesFM` (or gather information from other PR) to see whether is possible to bypass the internal pipeline and just rearrange the batch such as in chronos? THX
